libtiledbsoma/src/soma/soma_dataframe.h

```diff
@@ -0,0 +1,213 @@
+/**
+ * @file   soma_dataframe.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ *   This file defines the SOMADataFrame class.
+ */
+
+#ifndef SOMA_DATAFRAME
+#define SOMA_DATAFRAME
+
+#include <tiledb/tiledb>
+#include "soma_object.h"
+
+namespace tiledbsoma {
+
+class SOMAArray;
+class ArrayBuffers;
+
+using namespace tiledb;
+
+class SOMADataFrame : public SOMAObject {
+   public:
+    //===================================================================
+    //= public static
+    //===================================================================
+
+    /**
+     * @brief Create a SOMADataFrame object at the given URI.
+     *
+     * @param ctx TileDB context
+     * @param uri URI to create the SOMADataFrame
+     * @param schema TileDB ArraySchema
+     */
+    static void create(
+        std::shared_ptr<Context> ctx, std::string_view uri, ArraySchema schema);
+
+    /**
+     * @brief Open and return a SOMADataFrame object at the given URI.
+     *
+     * @param uri URI to create the SOMADataFrame
+     * @param schema TileDB ArraySchema
+     * @param platform_config Platform-specific options used to create this
+     * DataFrame
+     * @param column_names A list of column names to use as user-defined index
+     * columns (e.g., ``['cell_type', 'tissue_type']``). All named columns must
+     * exist in the schema, and at least one index column name is required.
+     * @param timestamp If specified, overrides the default timestamp used to
+     * open this object. If unset, uses the timestamp provided by the context.
+     * @return std::shared_ptr<SOMADataFrame> SOMADataFrame
+     */
+    static std::shared_ptr<SOMADataFrame> open(
```

Comment:
For a later time, all these `open` functions could return a `unique_ptr`, which can be moved to a `shared_ptr` if needed by the caller. But it doesn't have to be done now.